ambry-frontend/src/main/java/com.github.ambry.frontend/GetPeersHandler.java

```diff
@@ -0,0 +1,171 @@
+   * succeeded.
+   */
+  private class SecurityProcessRequestCallback implements Callback<Void> {
+    private final RestRequest restRequest;
+    private final RestResponseChannel restResponseChannel;
+    private final Callback<ReadableStreamChannel> callback;
+    private final long operationStartTimeMs;
+
+    SecurityProcessRequestCallback(RestRequest restRequest, RestResponseChannel restResponseChannel,
+        Callback<ReadableStreamChannel> callback) {
+      this.restRequest = restRequest;
+      this.restResponseChannel = restResponseChannel;
+      this.callback = callback;
+      operationStartTimeMs = SystemTime.getInstance().milliseconds();
+    }
+
+    /**
+     * If {@code exception} is null, gathers all the peers of the datanode that corresponds to the params in the request
+     * and returns them as a part of the response body.
+     * @param result The result of the request. This would be non null when the request executed successfully
+     * @param exception The exception that was reported on execution of the request
+     */
+    @Override
+    public void onCompletion(Void result, Exception exception) {
+      long processingStartTimeMs = SystemTime.getInstance().milliseconds();
+      metrics.getPeersSecurityProcessingTimeInMs.update(processingStartTimeMs - operationStartTimeMs);
+      try {
+        if (exception == null) {
+          DataNodeId dataNodeId = getDataNodeId(restRequest);
+          LOGGER.trace("Getting peer hosts for {}:{}", dataNodeId.getHostname(), dataNodeId.getPort());
+          Set<DataNodeId> peerDataNodeIds = new HashSet<>();
+          List<? extends ReplicaId> replicaIdList = clusterMap.getReplicaIds(dataNodeId);
+          for (ReplicaId replicaId : replicaIdList) {
+            List<? extends ReplicaId> peerReplicaIds = replicaId.getPeerReplicaIds();
+            for (ReplicaId peerReplicaId : peerReplicaIds) {
+              peerDataNodeIds.add(peerReplicaId.getDataNodeId());
+            }
+          }
+          StringBuilder peerSb = new StringBuilder();
+          for (DataNodeId peerDataNodeId : peerDataNodeIds) {
+            peerSb.append(peerDataNodeId.getHostname()).append(":").append(peerDataNodeId.getPort()).append(",");
+          }
+          if (peerSb.length() > 0) {
+            peerSb.deleteCharAt(peerSb.length() - 1);
+          }
+          String peerStr = peerSb.toString();
+          ReadableStreamChannel channel = new ByteBufferReadableStreamChannel(ByteBuffer.wrap(peerStr.getBytes()));
+          restResponseChannel.setHeader(RestUtils.Headers.CONTENT_TYPE, "text/plain");
+          restResponseChannel.setHeader(RestUtils.Headers.CONTENT_LENGTH, channel.getSize());
+          callback.onCompletion(channel, null);
+        } else {
+          callback.onCompletion(null, exception);
+        }
+      } catch (Exception e) {
+        callback.onCompletion(null, e);
```

Comment:
minor: i think callback.onCompletion can be called in the finally block. In the catch block we can just do `exception = e`. Then we can call the callback at one point in the function.